src/dftbp/dftbplus/apicallback.F90

```diff
@@ -0,0 +1,554 @@
+
+    !> Optional BLACS descriptor for the matrix in dataBuf. Not present if SCALAPACK is not
+    !! supported
+    integer, intent(in), target, optional :: blacsDescr(:)
+
+    if (.not. associated(this%dm_callback)) then
+      return
+    endif
+
+    call this%dm_callback(this%dmAuxPtr, iKpoint, iSpin, blacsDescr=blacsDescr, dataBufReal=dataBuf)
+
+  end subroutine TAPICallback_invokeDM_real
+
+
+  !> Callback for complex density matrix
+  subroutine TAPICallback_invokeDM_cplx(this, iKpoint, iSpin, dataBuf, blacsDescr)
+
+    !> Instance
+    class(TAPICallback) :: this
+
+    !> Index of k-point
+    integer, value :: iKpoint
+
+    !> Index of spin chanel
+    integer, value :: iSpin
+
+    !> Density matrix in dense format
+    complex(dp), intent(inout), target :: dataBuf(:,:)
+
+    !> Optional BLACS descriptor for the matrix in dataBuf. Not present if SCALAPACK is not
+    !! supported
+    integer, intent(in), target, optional :: blacsDescr(:)
+
+    if (.not. associated(this%dm_callback)) then
+      return
+    endif
+
+    call this%dm_callback(this%dmAuxPtr, iKpoint, iSpin, blacsDescr=blacsDescr, dataBufCplx=dataBuf)
+
+  end subroutine TAPICallback_invokeDM_cplx
+
+
+  !> Register an overlap matrix callback
+  subroutine TAPICallback_registerS(this, callback, auxPtr)
+
+    !> Instance
+    class(TAPICallback) :: this
+
+    !> Callback procedure to invoke
+    procedure(TDMHSCallbackFunc) :: callback
+
+    !> Auxillary data set by the callback
+    class(*), pointer :: auxPtr
+
+    this%s_callback => callback
+    this%sAuxPtr => auxPtr
+
+  end subroutine TAPICallback_registerS
+
+
+  !> Register an overlap matrix setting callback
+  subroutine TAPICallback_registerSetS(this, callback, auxPtr)
+
+    !> Instance
+    class(TAPICallback) :: this
+
+    !> Callback procedure to invoke
+    procedure(TSetDMHSCallbackFunc) :: callback
+
+    !> Auxillary data set by the callback
+    class(*), pointer :: auxPtr
+
+    this%set_s_callback => callback
+    this%set_sAuxPtr => auxPtr
+
+  end subroutine TAPICallback_registerSetS
+
+  !> Callback for real overlap matrix
+  subroutine TAPICallback_invokeS_real(this, iKpoint, iSpin, dataBuf, blacsDescr)
+
+    !> Instance
+    class(TAPICallback) :: this
+
+    !> Index of k-point
+    integer, value :: iKpoint
+
+    !> Index of spin chanel
+    integer, value :: iSpin
+
+    !> Overlap matrix in dense format
+    real(dp), intent(inout), target :: dataBuf(:,:)
+
+    !> Optional BLACS descriptor for the matrix in dataBuf. Not present if SCALAPACK is not
+    !! supported
+    integer, intent(in), target, optional :: blacsDescr(:)
+
+    integer res
+
+    if (associated(this%s_callback)) then
+      call this%s_callback(this%sAuxPtr, iKpoint, iSpin, blacsDescr=blacsDescr, dataBufReal=dataBuf)
+    endif
+
+    if (associated(this%set_s_callback)) then
+      res = this%set_s_callback(this%set_sAuxPtr, iKpoint, iSpin, blacsDescr=blacsDescr, dataBufReal=dataBuf)
```

Comment:
This isn't return value of set_s_callback isn't used anywhere on this side of the ASI interaction (likewise in other invoke routines later in this file). Whether the data has been changed should be propagated upwards.